tiny_cnn/lossfunctions/loss_function.h

```diff
@@ -32,48 +32,67 @@ namespace tiny_cnn {
 // mean-squared-error loss function for regression
 class mse {
 public:
-    static float_t f(float_t y, float_t t) {
-        return (y - t) * (y - t) / 2;
+    static float_t f(const vec_t& y, const vec_t& t) {
+        assert(y.size() == t.size());
+        float_t d = 0.0;
+        for(unsigned int i = 0; i < y.size(); ++i)
+            d += (y[i] - t[i]) * (y[i] - t[i]);
+        return d/y.size();
     }
 
```

Comment:
The change from `float f(float,float)` to `float f(vec_t,vec_t)` was necessary to support loss functions computed over a vector instead of accumulating the element wise loss (i.e. MSE where the loss is divided by the number of elements).

I've changed the df functions in respond to your comment on #217. I'm not aware of functions requiring the whole vector to compute the derivative. This change might not be necessary but might be more convenient as it doesn't require a extra loop where the derivative is used.

`std::accumulate` is only suitable for operations on one vector, to my knowledge.